rpc/src/rpc.rs

```diff
@@ -1861,36 +1861,59 @@ impl JsonRpcRequestProcessor {
                 "Invalid param: not a Token mint".to_string(),
             ));
         }
-        let mut token_balances: Vec<_> = self
-            .get_filtered_spl_token_accounts_by_mint(&bank, &mint_owner, mint, vec![])?
-            .into_iter()
-            .map(|(address, account)| {
-                let amount = StateWithExtensions::<TokenAccount>::unpack(account.data())
-                    .map(|account| account.base.amount)
-                    .unwrap_or(0);
-                (address, amount)
-            })
-            .collect();
 
-        let sort_largest = |a: &(_, u64), b: &(_, u64)| b.1.cmp(&a.1);
+        #[derive(Debug, PartialEq, Eq)]
+        struct TokenAccountBalance {
+            address: Pubkey,
+            amount: u64,
+        }
+        impl PartialOrd for TokenAccountBalance {
+            fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+                Some(self.amount.cmp(&other.amount).reverse())
+            }
+        }
+        impl Ord for TokenAccountBalance {
+            fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+                self.amount.cmp(&other.amount).reverse()
+            }
+        }
 
-        let largest_token_balances = if token_balances.len() > NUM_LARGEST_ACCOUNTS {
-            token_balances
-                .select_nth_unstable_by(NUM_LARGEST_ACCOUNTS, sort_largest)
-                .0
-        } else {
-            token_balances.as_mut_slice()
-        };
-        largest_token_balances.sort_unstable_by(sort_largest);
+        let mut token_balances =
+            BinaryHeap::<TokenAccountBalance>::with_capacity(NUM_LARGEST_ACCOUNTS);
+        for (address, account) in
+            self.get_filtered_spl_token_accounts_by_mint(&bank, &mint_owner, mint, vec![])?
+        {
+            let amount = StateWithExtensions::<TokenAccount>::unpack(account.data())
+                .map(|account| account.base.amount)
+                .unwrap_or(0);
 
-        let largest_token_balances = largest_token_balances
-            .iter()
-            .map(|(address, amount)| RpcTokenAccountBalance {
-                address: address.to_string(),
-                amount: token_amount_to_ui_amount(*amount, decimals),
+            let requires_insert = if token_balances.len() < NUM_LARGEST_ACCOUNTS {
+                true
+            } else {
+                if let Some(peek) = token_balances.peek() {
+                    peek.amount < amount
```

Comment:
This logic will potentially return different lists if there are multiple accounts with the same balance at the `NUM_LARGEST_ACCOUNTS` boundary, depending on what order the node happens to scan accounts (and whether the node is using secondary indexes). Since we're going to the trouble of doing this, I'd prefer to emulate the code for largest native accounts:
https://github.com/solana-labs/solana/blob/bf2e8ee32f2280692f94f1f21d6fe2cb82700852/accounts-db/src/accounts.rs#L251-L260
You might be able to rework your Ord/PartialOrd definitions, but honestly it's probably simplest to just use a tuple as that code I linked does.

Reply:
Awesome! Looks much better! And yeah, I missed the case that for the same amount we can get different addresses :sob: